File: src/main/java/io/github/chindeaone/collectiontracker/gui/OverlayManager.java
package io.github.chindeaone.collectiontracker.gui;

import io.github.chindeaone.collectiontracker.gui.overlays.*;

import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.Map;

public class OverlayManager {

    private static final Map<String, AbstractOverlay> overlays = new LinkedHashMap<>();
    private static boolean globalRenderingAllowed = true;

    public static void add(AbstractOverlay overlay) {
        overlays.put(overlay.overlayLabel(), overlay);
    }

    public static Collection<AbstractOverlay> all() {
        return overlays.values();
    }

    public static void setGlobalRendering(boolean allowed) {
        globalRenderingAllowed = allowed;
        overlays.values().forEach(o -> o.setRenderingAllowed(allowed));
    }

    public static boolean isInEditorMode() {
        return !globalRenderingAllowed;
    }

    public static void setTrackingOverlayRendering(boolean allowed) {
        AbstractOverlay overlay = overlays.get("Collection Overlay");
        if (overlay != null) {
            overlay.setRenderingAllowed(allowed);
        }
    }

    public static void setSkillOverlayRendering(boolean allowed) {
        AbstractOverlay overlay = overlays.get("Skill Overlay");
        if (overlay != null) {
            overlay.setRenderingAllowed(allowed);
        }
    }

    public static void overlayRegistration() {
        OverlayManager.add(new CollectionOverlay());
        OverlayManager.add(new MiningStatsOverlay());
        OverlayManager.add(new CommissionsOverlay());
        OverlayManager.add(new SkillOverlay());
        OverlayManager.add(new ForagingStatsOverlay());
        OverlayManager.add(new SkyMallOverlay());
        OverlayManager.add(new LotteryOverlay());
        OverlayManager.add(new PickaxeAbilityOverlay());
        OverlayManager.add(new AxeAbilityOverlay());
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/OverlayManager.kt
package io.github.chindeaone.collectiontracker.gui

import io.github.chindeaone.collectiontracker.gui.overlays.*

object OverlayManager {

    private val overlays: MutableMap<String, AbstractOverlay> = linkedMapOf()
    private var globalRenderingAllowed: Boolean = true

    fun add(overlay: AbstractOverlay) {
        overlays[overlay.overlayLabel()] = overlay
    }

    @JvmStatic
    fun all(): Collection<AbstractOverlay> = overlays.values

    fun setGlobalRendering(allowed: Boolean) {
        globalRenderingAllowed = allowed
        overlays.values.forEach { it.isRenderingAllowed = allowed }
    }

    fun isInEditorMode(): Boolean = !globalRenderingAllowed

    private fun setOverlayRendering(label: String, allowed: Boolean) {
        overlays[label]?.isRenderingAllowed = allowed
    }

    @JvmStatic
    fun setTrackingOverlayRendering(allowed: Boolean) = setOverlayRendering(CollectionOverlay().overlayLabel(), allowed)

    fun setMultiTrackingOverlayRendering(allowed: Boolean) = setOverlayRendering(MultiCollectionOverlay().overlayLabel(), allowed)

    @JvmStatic
    fun setSkillOverlayRendering(allowed: Boolean) = setOverlayRendering(SkillOverlay().overlayLabel(), allowed)

    fun setColeweightOverlayRendering(allowed: Boolean) = setOverlayRendering(ColeweightOverlay().overlayLabel(), allowed)

    fun overlayRegistration() {
        listOf(
                CollectionOverlay(),
                MultiCollectionOverlay(),
                MiningStatsOverlay(),
                CommissionsOverlay(),
                SkillOverlay(),
                ForagingStatsOverlay(),
                SkyMallOverlay(),
                LotteryOverlay(),
                PickaxeAbilityOverlay(),
                AxeAbilityOverlay(),
                DeployableOverlay(),
                TemporaryBuffsOverlay(),
                TitleOverlay(),
                TimerOverlay(),
                StopwatchOverlay(),
                ColeweightOverlay()
        ).forEach { add(it) }
    }

    @JvmStatic
    fun isCollectionOverlay(overlay: AbstractOverlay): Boolean {
        val label = overlay.overlayLabel()
        return label == CollectionOverlay().overlayLabel() || label == MultiCollectionOverlay().overlayLabel()
    }

    @JvmStatic
    fun getTimerOverlay(): TimerOverlay? = overlays.values.filterIsInstance<TimerOverlay>().firstOrNull()

    @JvmStatic
    fun getStopwatchOverlay(): StopwatchOverlay? = overlays.values.filterIsInstance<StopwatchOverlay>().firstOrNull()
}

